internal/service/subscription_state_handler.go

```diff
@@ -0,0 +1,84 @@
+package service
+
+import (
+	"github.com/flexprice/flexprice/internal/domain/creditgrant"
+	"github.com/flexprice/flexprice/internal/domain/subscription"
+	ierr "github.com/flexprice/flexprice/internal/errors"
+	"github.com/flexprice/flexprice/internal/types"
+)
+
+type StateAction string
+
+const (
+	StateActionApply  StateAction = "apply"
+	StateActionSkip   StateAction = "skip"
+	StateActionDefer  StateAction = "defer"
+	StateActionCancel StateAction = "cancel"
+)
+
+func NewSubscriptionStateHandler(subscription *subscription.Subscription, grant *creditgrant.CreditGrant) *SubscriptionStateHandler {
+	return &SubscriptionStateHandler{
+		subscription: subscription,
+		grant:        grant,
+	}
+}
+
+type SubscriptionStateHandler struct {
+	subscription *subscription.Subscription
+	grant        *creditgrant.CreditGrant
+}
+
+func (h *SubscriptionStateHandler) DetermineAction() (StateAction, error) {
+	switch h.subscription.SubscriptionStatus {
+	case types.SubscriptionStatusActive:
+		return StateActionApply, nil
+
+	case types.SubscriptionStatusTrialing:
+		// Check if grant should apply during trial period
+		if h.shouldApplyDuringTrial() {
+			return StateActionApply, nil
+		}
+		return StateActionDefer, nil
+
+	case types.SubscriptionStatusPastDue:
+		// For past due, we defer rather than skip to allow recovery
+		return StateActionDefer, nil
+
+	case types.SubscriptionStatusUnpaid:
+		// Similar to past due - defer for potential recovery
+		return StateActionDefer, nil
+
+	case types.SubscriptionStatusCancelled:
+		// Cancelled subscriptions should not receive new credits
+		return StateActionCancel, nil
+
+	case types.SubscriptionStatusIncomplete:
+		// Defer incomplete subscriptions as they might become active
+		return StateActionDefer, nil
+
+	case types.SubscriptionStatusIncompleteExpired:
+		// Expired incomplete subscriptions should be cancelled
+		return StateActionCancel, nil
+
+	case types.SubscriptionStatusPaused:
+		// Paused subscriptions should defer credits until resumed
+		return StateActionDefer, nil
```

Comment:
TODO: this should be skip - so we mark the current grant skipped and we schedule the next one
And from what I am taking away for defer - we will basically wait to decide when to process but we have an intent to process ex say payment is awaited, so we are deferring and we will process once the subscription is paid.
In future we might start using skip more than deferred.